LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClientExtensions.cs

```diff
@@ -0,0 +1,120 @@
+        }
+
+        private sealed class ResilientClient : ILoRaDeviceClient, IIdentityProvider<ILoRaDeviceClient>
+        {
+            private const int MaxAttempts = 3;
+
+            private readonly ILoRaDeviceClient client;
+            private readonly ILogger? logger;
+
+            public ResilientClient(ILoRaDeviceClient client, ILogger<ResilientClient>? logger)
+            {
+                this.client = client;
+                this.logger = logger;
+            }
+
+            public bool EnsureConnected() => this.client.EnsureConnected();
+            public Task DisconnectAsync(CancellationToken cancellationToken) => this.client.DisconnectAsync(cancellationToken);
+            ValueTask IAsyncDisposable.DisposeAsync() => this.client.DisposeAsync();
+
+            private sealed class Void
+            {
+                public static readonly Void Value = new();
+                private Void() { }
+            }
+
+            private Task<TResult> InvokeAsync<TResult>(Operation<Void, Void, TResult> operation,
+                                                       CancellationToken cancellationToken,
+                                                       [CallerMemberName] string? operationName = null) =>
+                InvokeAsync(operation, Void.Value, Void.Value, cancellationToken, operationName);
+
+            private Task<TResult> InvokeAsync<T, TResult>(Operation<T, Void, TResult> operation, T arg,
+                                                          CancellationToken cancellationToken,
+                                                          [CallerMemberName] string? operationName = null) =>
+                InvokeAsync(operation, arg, Void.Value, cancellationToken, operationName);
+
+            private async Task<TResult> InvokeAsync<T1, T2, TResult>(Operation<T1, T2, TResult> operation, T1 arg1, T2 arg2,
+                                                                     CancellationToken cancellationToken,
+                                                                     [CallerMemberName] string? operationName = null)
+            {
+                for (var attempt = 1; ; attempt++)
+                {
+                    try
+                    {
+                        _ = this.client.EnsureConnected();
+                        return await operation(this.client, arg1, arg2, cancellationToken);
+                    }
+                    catch (Exception ex)
+                        when ((ex is ObjectDisposedException
+                                  || (ex is InvalidOperationException ioe
+                                      && ioe.Message.StartsWith("This operation is only allowed using a successfully authenticated context.", StringComparison.OrdinalIgnoreCase)))
+                              && ExceptionFilterUtility.True(() => this.logger?.LogWarning(ex, @"Device client operation ""{Operation}"" failed due to: {Error}", operationName, ex.GetBaseException().Message)))
```

Comment:
Can you please explicitely state here that the operation will be retryied?

Reply:
Added with 8babf8a84e80d52ebf2bd225cb90d9d4fd882cdd.